Using the output of a Gradle task to set the value of a property on `FlankGradleExtension`

I'm trying to set the value of the `testTargets` property using the ouput of a task. It looks like the `testTargets` property is getting resolved at configuration time, leading to a `java.io.FileNotFoundException` (it tries to read a file that would be generated by the task).

Is this a bug, or something that isn't generally supported?

### Context

One feature we’ve implemented in our current non-fladle flank setup is a “stability gate” for new and changed tests — we use `git diff` to generate a list of test classes that have been modified or added in a pull request, and then dynamically generate a flank yml config that runs those tests a large number of times to check if any flakiness was introduced.

I’m trying to implement this functionality when using fladle. My first thought was to wrap the git invocation in a gradle task, and wire up an output property from the task to the `testTargets` property in the fladle extension:

```kotlin
// In app/build.gradle.kts

open class GetChangedTestTargets : DefaultTask() {
  @OutputFile
  val outputFile: RegularFileProperty = project.objects.fileProperty()

  @TaskAction
  fun action() {
    // TODO(fladle): execute a git diff between the current branch and master to calculate which test targets should run.
    outputFile.asFile.get().writeText(
      """
      class foo.bar.SomeTestClass1
      class foo.bar.SomeTestClass2
      """.trimIndent()
    )
  }
}

val stabilityGateTestTargets by tasks.registering(GetChangedTestTargets::class) {
  // This task should never be considered up to date because it depends on state outside of the build: it invokes git.
  outputs.upToDateWhen { false }
  outputFile.set(layout.buildDirectory.file("stabilityGateTestTargets.txt"))
}

fladle {
  configs {
    // ...
    create("stabilityGateTests") {
      // ...
      testTargets.set(stabilityGateTestTargets.map { it.outputFile.get().asFile.readLines() })
    }
  }
}
```

… but unfortunately it appears that the testTargets is getting resolved at configuration time, because I get:
```
A problem occurred configuring project ':app'.
> java.io.FileNotFoundException: /Users/jschear/dev/myproject/app/build/stabilityGateTestTargets.txt (No such file or directory)
```

Looking at the stacktrace, it looks like the `validateOptionsUsed` function is the culprit.
```
Caused by: org.gradle.api.UncheckedIOException: java.io.FileNotFoundException: /Users/jschear/dev/myproject/app/build/stabilityGateTestTargets.txt (No such file or directory)
        at org.gradle.internal.UncheckedException.throwAsUncheckedException(UncheckedException.java:61)
        at org.gradle.internal.UncheckedException.throwAsUncheckedException(UncheckedException.java:41)
        at org.gradle.api.internal.provider.AbstractProperty.calculatePresence(AbstractProperty.java:63)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.isPresent(AbstractMinimalProvider.java:74)
        at com.osacky.flank.gradle.FladleConfig$DefaultImpls.getPresentProperties(FladleConfig.kt:462)
        at com.osacky.flank.gradle.FladleConfigImpl.getPresentProperties(FladleConfigImpl.kt:9)
        at com.osacky.flank.gradle.validation.ValidateOptionsKt.validateOptionsUsed(ValidateOptions.kt:8)
        at com.osacky.flank.gradle.FladlePluginDelegate.createTasksForConfig(FladlePluginDelegate.kt:72)
        at com.osacky.flank.gradle.FladlePluginDelegate.access$createTasksForConfig(FladlePluginDelegate.kt:15)
        at com.osacky.flank.gradle.FladlePluginDelegate$configureTasks$1.execute(FladlePluginDelegate.kt:64)
        at com.osacky.flank.gradle.FladlePluginDelegate$configureTasks$1.execute(FladlePluginDelegate.kt:15)
```

--- a/fladle-plugin/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt
+++ b/fladle-plugin/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt
@@ -3,6 +3,7 @@ package com.osacky.flank.gradle
 import com.android.build.gradle.AppExtension
 import com.android.build.gradle.TestedExtension
 import com.android.build.gradle.api.BaseVariant
+import com.android.build.gradle.internal.tasks.factory.dependsOn
 import com.android.builder.model.TestOptions
 import com.osacky.flank.gradle.validation.checkForExclusionUsage
 import com.osacky.flank.gradle.validation.validateOptionsUsed
@@ -71,13 +72,23 @@ class FladlePluginDelegate {
       }
     }
 
-    checkIfSanityAndValidateConfigs(config)
-    validateOptionsUsed(config = config, flank = base.flankVersion.get())
-    checkForExclusionUsage(config)
     val configName = name.toLowerCase()
     // we want to use default dir only if user did not set own `localResultsDir`
     val useDefaultDir = config.localResultsDir.isPresent.not()
+
+    val validateFladle = register("validateFladleConfig$name") {
+      description = "Perform validation actions"
+      group = TASK_GROUP
+      doLast {
+        checkIfSanityAndValidateConfigs(config)
+        validateOptionsUsed(config = config, flank = base.flankVersion.get())
+        checkForExclusionUsage(config)
+      }
+    }
+
     val writeConfigProps = register("writeConfigProps$name", YamlConfigWriterTask::class.java, base, config, name)
+
+    writeConfigProps.dependsOn(validateFladle)
 
     register("printYml$name") {
       description = "Print the flank.yml file to the console."

--- a/fladle-plugin/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt
+++ b/fladle-plugin/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt
@@ -53,6 +53,14 @@ class FladlePluginDelegate {
       // Must be done afterEvaluate otherwise extension values will not be set.
       project.dependencies.add(FLADLE_CONFIG, "${base.flankCoordinates.get()}:${base.flankVersion.get()}")
 
+      // Only use automatic apk path detection for 'com.android.application' projects.
+      project.pluginManager.withPlugin("com.android.application") {
+        // This doesn't work properly for multiple configs since they likely are inheriting the config from root already. See #60 https://github.com/runningcode/fladle/issues/60
+        if (!base.debugApk.isPresent || !base.instrumentationApk.isPresent) {
+          findDebugAndInstrumentationApk(project, base)
+        }
+      }
+
       tasks.apply {
         createTasksForConfig(base, base, project, "")
 
@@ -64,13 +72,6 @@ class FladlePluginDelegate {
   }
 
   private fun TaskContainer.createTasksForConfig(base: FlankGradleExtension, config: FladleConfig, project: Project, name: String) {
-    // Only use automatic apk path detection for 'com.android.application' projects.
-    project.pluginManager.withPlugin("com.android.application") {
-      // This doesn't work properly for multiple configs since they likely are inheriting the config from root already. See #60 https://github.com/runningcode/fladle/issues/60
-      if (!config.debugApk.isPresent || !config.instrumentationApk.isPresent) {
-        findDebugAndInstrumentationApk(project, config)
-      }
-    }
 
     val configName = name.toLowerCase()
     // we want to use default dir only if user did not set own `localResultsDir`

--- a/fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/FlankAuthTestTask.kt
+++ b/fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/FlankAuthTestTask.kt
@@ -1,6 +1,7 @@
 package com.osacky.flank.gradle.integration
 
 import com.google.common.truth.Truth.assertThat
+import org.junit.Ignore
 import org.junit.Rule
 import org.junit.Test
 import org.junit.rules.TemporaryFolder
@@ -9,6 +10,7 @@ class FlankAuthTestTask {
   @get:Rule
   var testProjectRoot = TemporaryFolder()
 
+  @Ignore("Flaky. See https://github.com/runningcode/fladle/issues/245")
   @Test
   fun testFlankAuth() {
     // We set a task timeout because running flankAuth opens a link in the web browser.

--- a/fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/FlankGradlePluginIntegrationTest.kt
+++ b/fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/FlankGradlePluginIntegrationTest.kt
@@ -157,7 +157,7 @@ class FlankGradlePluginIntegrationTest {
             id "com.osacky.fladle"
            }
            fladle {
-             serviceAccountCredentials = project.layout.projectDirectory.file("foo")
+             serviceAccountCredentials = project.layout.projectDirectory.file("flank-gradle-service.json")
              debugApk = "test-debug.apk"
              instrumentationApk = "instrumentation-debug.apk"
              roboScript = "foo.script"
@@ -180,7 +180,7 @@ class FlankGradlePluginIntegrationTest {
            id "com.osacky.fladle"
          }
          fladle {
-           serviceAccountCredentials = project.layout.projectDirectory.file("foo")
+           serviceAccountCredentials = project.layout.projectDirectory.file("flank-gradle-service.json")
            debugApk = "test-debug.apk"
            instrumentationApk = "instrumentation-debug.apk"
            configs {
@@ -190,7 +190,7 @@ class FlankGradlePluginIntegrationTest {
          }
          """.trimMargin()
     )
-    testProjectRoot.newFile("foo").writeText("{}")
+    testProjectRoot.writeEmptyServiceCredential()
     val result = testProjectRoot.gradleRunner()
       .withGradleVersion("7.0-rc-1")
       .withArguments("printYmlFooConfig")

--- a/fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/FlankGradlePluginIntegrationTest.kt
+++ b/fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/FlankGradlePluginIntegrationTest.kt
@@ -16,8 +16,7 @@ class FlankGradlePluginIntegrationTest {
   val oldVersion = "5.3.1"
 
   fun writeBuildGradle(build: String) {
-    val file = testProjectRoot.newFile("build.gradle")
-    file.writeText(build)
+    testProjectRoot.writeBuildDotGradle(build)
   }
 
   @Test

--- a/fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/SanityRoboTest.kt
+++ b/fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/SanityRoboTest.kt
@@ -31,7 +31,7 @@ class SanityRoboTest {
       |  instrumentationApk = "test.apk"
       |  configs {
       |    sanity {
-      |      makeSanityRun()
+      |      clearPropertiesForSanityRobo()
       |    }
       |  }
       |}
@@ -44,10 +44,9 @@ class SanityRoboTest {
     assertThat(result.output).contains("FAILED")
     assertThat(result.output).contains(baseConfigMessage("instrumentationApk"))
 
-    val resultSanity = runner.withArguments("printYmlSanity").buildAndFail()
+    val resultSanity = runner.withArguments("printYmlSanity").build()
 
-    assertThat(resultSanity.output).contains("FAILED")
-    assertThat(result.output).contains(baseConfigMessage("instrumentationApk"))
+    assertThat(resultSanity.output).contains("SUCCESS")
   }
 
   @Test
@@ -159,10 +158,9 @@ class SanityRoboTest {
     val expectedMessage = additionalConfigMessage("roboScript", "sanity")
 
     val runner = testProjectRoot.gradleRunner()
-    val result = runner.withArguments("printYml").buildAndFail()
+    val result = runner.withArguments("printYml").build()
 
-    assertThat(result.output).contains("FAILED")
-    assertThat(result.output).contains(expectedMessage)
+    assertThat(result.output).contains("SUCCESS")
 
     val resultOrange = runner.withArguments("printYmlSanity").buildAndFail()
 

--- a/fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/TestFixtures.kt
+++ b/fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/TestFixtures.kt
@@ -19,5 +19,5 @@ fun TemporaryFolder.gradleRunner() =
     .withProjectDir(root)
 
 fun TemporaryFolder.writeEmptyServiceCredential() {
-  newFile("foo").writeText("{}")
+  newFile("flank-gradle-service.json").writeText("{}")
 }

--- a/fladle-plugin/src/test/java/com/osacky/flank/gradle/validation/ValidateExclusionsTest.kt
+++ b/fladle-plugin/src/test/java/com/osacky/flank/gradle/validation/ValidateExclusionsTest.kt
@@ -3,6 +3,7 @@ package com.osacky.flank.gradle.validation
 import com.google.common.truth.Truth.assertThat
 import com.osacky.flank.gradle.integration.gradleRunner
 import com.osacky.flank.gradle.integration.writeBuildDotGradle
+import com.osacky.flank.gradle.integration.writeEmptyServiceCredential
 import org.junit.Rule
 import org.junit.Test
 import org.junit.rules.TemporaryFolder
@@ -61,11 +62,11 @@ class ValidateExclusionsTest {
     """.trimMargin()
     )
 
+    testProjectRoot.writeEmptyServiceCredential()
     val runner = testProjectRoot.gradleRunner()
 
-    runner.withArguments("printYml").buildAndFail().run {
-      assertThat(output).contains("FAILED")
-      assertThat(output).contains("Options testShards and maxTestShards cannot be used together. Choose one of them.")
+    runner.withArguments("printYml").build().run {
+      assertThat(output).contains("SUCCESS")
     }
 
     runner.withArguments("printYmlNewSharding").buildAndFail().run {

--- a/fladle-plugin/src/test/java/com/osacky/flank/gradle/validation/ValidateOptionsTest.kt
+++ b/fladle-plugin/src/test/java/com/osacky/flank/gradle/validation/ValidateOptionsTest.kt
@@ -84,9 +84,8 @@ class ValidateOptionsTest {
 
     val runner = testProjectRoot.gradleRunner()
 
-    runner.withArguments("printYml").buildAndFail().run {
-      assertThat(output).contains("FAILED")
-      assertThat(output).contains("Option useAverageTestTimeForNewTests is available since flank 20.8.4, which is higher than used 20.5.0")
+    runner.withArguments("printYml").build().run {
+      assertThat(output).contains("SUCCESS")
     }
 
     runner.withArguments("printYmlNewNetwork").buildAndFail().run {

--- a/fladle-plugin/src/test/java/com/osacky/flank/gradle/validation/ValidateOptionsTest.kt
+++ b/fladle-plugin/src/test/java/com/osacky/flank/gradle/validation/ValidateOptionsTest.kt
@@ -93,9 +93,8 @@ class ValidateOptionsTest {
       assertThat(output).contains("Option useAverageTestTimeForNewTests is available since flank 20.8.4, which is higher than used 20.5.0")
     }
 
-    runner.withArguments("printYmlNoSharding").buildAndFail().run {
-      assertThat(output).contains("FAILED")
-      assertThat(output).contains("Option useAverageTestTimeForNewTests is available since flank 20.8.4, which is higher than used 20.5.0")
+    runner.withArguments("printYmlNoSharding").build().run {
+      assertThat(output).contains("SUCCESS")
     }
   }
 

